### ISSUE TITLE
How can I define theorem envrionment with macros?

### ISSUE THREAD

**Issue by fgborges** [michael-brade/LaTeX.js#64]
Hi Michael,

Thank you for a great software, I'm really interested in this software for displaying my academic works. my papers is written with many theorem and that kind of environments. I guess I can add a macro which simulate the behavior of that environment.

You have already had an example to define macros but, I'm sorry for that, I couldn't define that macro. Could you give us more examples for defining macros?

Firstly, If you give a few ones to define theorems. Then I may contribute your project by adding more examples to your documents.

Thanks.

**Comment by fgborges** [comment]
Finally, I solved it by myself.
If you don't mind, I'd like to prose to add this example to your repository.
How about it?
```html
<!DOCTYPE html>
<html lang="en-US">
    <head>
        <meta charset="utf-8">
        <title>LaTeX.js Macros</title>
        <script src="https://unpkg.com/latex.js/dist/latex.min.js"></script>
        <style>
            .theorem, .proof {
                margin-top: 0.8rem;
            }
            .theorem > p:first-of-type, .proof > p:first-of-type {
                display: inline;
            }
            .proof > p:last-of-type::after {
                content: "\2610"
            }
        </style>
    </head>
    <body>
        <template id="latex-js">
            \documentclass{article}
            \begin{document}
            \section{Section}
            \begin{theorem}[Hola!]
            Hello!
            Good noon!
            \end{theorem}
            \begin{theorem}[Hola!]
            Hello!
            Good noon!
            \end{theorem}
            \begin{lemma}
            This is lemma.
            \end{lemma}
            \begin{proof}
            This is a proof.

            This is also a proof.
            \end{proof}
            \end{document}
        </template>
        <script>
            function html(strings, ...values) {
                const template = document.createElement('template');
                template.innerHTML = String.raw(strings, ...values).trim();
                return template.content.firstChild;
            }
            function CustomMacros(generator) {
                this.g = generator;
                this.g.newCounter('theorem');
            }
            const names = ['theorem', 'axiom', 'lemma', 'example', 'definition', 'remark'];
            CustomMacros.args = {
                ...Object.fromEntries(names.map(name=>[name, ['V', 'o?']])),
                proof: ['V', 'o?']
            };
            CustomMacros.prototype = {
                ...CustomMacros.prototype,
                ...names.map(name=>({
                    [name](option) {
                        this.g.startlist();
                        this.g.stepCounter('theorem');
                        const capitalized = name[0].toUpperCase()+name.slice(1);
                        const number = (this.g.counter('section') === 0 ? '' : this.g.counter('section')+'.')
                            + (this.g.counter('subsection') === 0 ? '' : this.g.counter('subsection')+'.')
                            + (this.g.counter('subsubsection') === 0 ? '' : this.g.counter('subsubsection')+'.')
                            + this.g.counter('theorem')
                        const theoremtitle = option
                            ? html`<span class="up"><span class="bf">${capitalized} ${number}</span> (${option.textContent}). </span>`
                            : html`<span class="up"><span class="bf">${capitalized} ${number}</span>. </span>`
                        return [this.g.create(this.g.block, [theoremtitle], 'theorem it')];
                    },
                    ['end'+name]() {
                        this.g.endlist();
                    }
                })).reduce(Object.assign),
                proof(option) {
                    this.g.startlist();
                    const prooftitle = option
                        ? html`<span class="it">${option.textContent}. </span>`
                        : html`<span class="it">Proof. </span>`;
                    return [this.g.create(this.g.block, [prooftitle], 'proof')];
                },
                endproof() {
                    this.g.endlist();
                }
            };
            const generator = new latexjs.HtmlGenerator({ hyphenate: false, CustomMacros });
            const text = document.getElementById("latex-js").innerHTML;
            const parsed = latexjs.parse(text, { generator })
            document.body.appendChild(parsed.stylesAndScripts('https://unpkg.com/latex.js/dist/'));
            document.body.appendChild(parsed.domFragment());
        </script>
    </body>
</html>
```

**Comment by michael-brade** [comment]
Sorry for being late and of no help in this case, too much going on! Your code looks really good, I wonder where you got stuck originally ;-)

> If you don't mind, I'd like to prose to add this example to your repository.

Absolutely! What did you have in mind: add it to README.md, or create a new folder `examples` in the repository? There is also `test/api`, but I'm not sure how many people are actually looking at it...

**Comment by fgborges** [comment]
Thanks! I'll add this to 'examples' folder. BTW, I also want to add 'theorem*' macro like this, but it seems that the parser rejects the name 'theorem*'. Why couldn't the library parse this?

```html
<!DOCTYPE html>
<html lang="en-US">
    <head>
        <meta charset="utf-8">
        <title>LaTeX.js Macros</title>
        <script src="https://unpkg.com/latex.js/dist/latex.min.js"></script>
        <style>
            .theorem, .proof {
                margin-top: 0.8rem;
            }
            .theorem > p:first-of-type, .proof > p:first-of-type {
                display: inline;
            }
            .proof > p:last-of-type::after {
                content: "\2610"
            }
        </style>
    </head>
    <body>
        <template id="latex-js">
            \documentclass{article}
            \begin{document}
            \section{Section}
            \begin{theorem}[Foo]
            Hello!
            \end{theorem}
            \begin{lemma*}
            This is lemma.
            \end{lemma*}
            \begin{proof}
            This is a proof.
            \end{proof}
            \end{document}
        </template>
        <script>
            function html(strings, ...values) {
                const template = document.createElement('template');
                template.innerHTML = String.raw(strings, ...values).trim();
                return template.content.firstChild;
            }
            function CustomMacros(generator) {
                this.g = generator;
                this.g.newCounter('theorem');
            }
            const names = ['theorem', 'axiom', 'lemma', 'example', 'definition', 'remark'];
            CustomMacros.args = {
                ...Object.fromEntries(names.map(name=>[name, ['V', 'o?']])),
                proof: ['V', 'o?']
            };
            CustomMacros.prototype = {
                ...CustomMacros.prototype,
                ...names.map(name=>({
                    [name+'*'](option) {
                        this.g.startlist();
                        const capitalized = name[0].toUpperCase()+name.slice(1);
                        const theoremtitle = option
                            ? html`<span class="up"><span class="bf">${capitalized}</span> (${option.textContent}). </span>`
                            : html`<span class="up"><span class="bf">${capitalized}</span>. </span>`
                        return [this.g.create(this.g.block, [theoremtitle], 'theorem it')];
                    },
                    ['end'+name+'*']() {
                        this.g.endlist();
                    },
                    [name](option) {
                        this.g.startlist();
                        this.g.stepCounter('theorem');
                        const capitalized = name[0].toUpperCase()+name.slice(1);
                        const number = (this.g.counter('section') === 0 ? '' : this.g.counter('section')+'.')
                            + (this.g.counter('subsection') === 0 ? '' : this.g.counter('subsection')+'.')
                            + (this.g.counter('subsubsection') === 0 ? '' : this.g.counter('subsubsection')+'.')
                            + this.g.counter('theorem')
                        const theoremtitle = option
                            ? html`<span class="up"><span class="bf">${capitalized} ${number}</span> (${option.textContent}). </span>`
                            : html`<span class="up"><span class="bf">${capitalized} ${number}</span>. </span>`
                        return [this.g.create(this.g.block, [theoremtitle], 'theorem it')];
                    },
                    ['end'+name]() {
                        this.g.endlist();
                    }
                })).reduce(Object.assign),
                proof(option) {
                    this.g.startlist();
                    const prooftitle = option
                        ? html`<span class="it">${option.textContent}. </span>`
                        : html`<span class="it">Proof. </span>`;
                    return [this.g.create(this.g.block, [prooftitle], 'proof')];
                },
                endproof() {
                    this.g.endlist();
                }
            };
            const generator = new latexjs.HtmlGenerator({ hyphenate: false, CustomMacros });
            const text = document.getElementById("latex-js").innerHTML;
            const parsed = latexjs.parse(text, { generator })
            document.body.appendChild(parsed.stylesAndScripts('https://unpkg.com/latex.js/dist/'));
            document.body.appendChild(parsed.domFragment());
        </script>
    </body>
</html>
```

**Comment by michael-brade** [comment]
Hi Fernando,
oh... I see. It makes sense. The problem is that `*` is not part of the `identifier` rule, which is what is expected as an environment name. I'll have to extend `identifier` or create a new rule.... I'll think about it. Thanks for finding this :-)

**Comment by fgborges** [comment]
Okay, you are going to fix identifier, aren't you? It's nice for the rule.
Temporary, I submit a pull request without `theorem*` and will do it again when you update the rule.

**Comment by michael-brade** [comment]
Darn... it's not as easy as it seemed. The star is not part of the macro name but an argument (kind of like `xparse` is doing it). But with environments the argument parsing is done only after the closing brace. I don't know a good solution yet. Maybe I'll have to add some special parsing for environments...

**Comment by michael-brade** [comment]
Hi Fernando, please try `theorem*` again, should work now :-) You'll have to define a macro that takes a star argument, like this:
```
    args.\theorem =      <[ V s ]>
    \theorem             : (s) ->
```

**Comment by fgborges** [comment]
Awesome! you're so cool. let me try it soon :)
BTW, I'll give you new pull request to combine all resource into one by data:url today.
Cheers 🍺
